### PR TITLE
[pfs] Expose PFS (Prompt Flow Service)

### PR DESCRIPTION
**Pull request by lalala123123**
# Description
1. export pfs to promptflow cli, pfs [-p PORT]
2. if not provide port, use random port
3. only current user can access the service.
4. only allow request from localhost

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/_constants.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
 FLOW_TOOLS_JSON_GEN_TIMEOUT = 60
 PROMPT_FLOW_DIR_NAME = ".promptflow"
 HOME_PROMPT_FLOW_DIR = (Path.home() / PROMPT_FLOW_DIR_NAME).resolve()
+SERVICE_CONFIG_FILE = "pf.yaml"
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_constants.py:22]: we should already have a constant for this? as this is sharing with authoring scenario?

File: src/promptflow/promptflow/_sdk/_service/entry.py [hunk 2]
```diff
@@ -1,16 +1,45 @@
...
+
+    parser.add_argument("-p", "--port", type=int, help="port of the promptflow service")
+    args = parser.parse_args(command_args)
+    port = args.port
+
+    if port and is_port_in_use():
+        raise UserErrorException(f"Service port {port} is used.")
+    if not port:
+        # Read and write permission for user.
+        mode = stat.S_IRUSR | stat.S_IWUSR
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_service/entry.py:34]: please reuse the logic in configuration part

File: src/promptflow/promptflow/_sdk/_service/entry.py [hunk 3]
```diff
@@ -1,16 +1,45 @@
...
+        (HOME_PROMPT_FLOW_DIR / SERVICE_CONFIG_FILE).touch(mode=mode, exist_ok=True)
+        with open(HOME_PROMPT_FLOW_DIR / SERVICE_CONFIG_FILE, "r") as f:
+            service_config = yaml.safe_load(f)
+            port = service_config.get("service", {}).get("port", None)
+
     app = create_app()
-    waitress.serve(app, port=LOCAL_SERVICE_PORT)
+    # Set host to localhost, only allow request from localhost.
+    waitress(app, host="127.0.0.1", port=port or get_random_port())
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_service/entry.py:42]: when do we write back the random port to config yaml？

File: src/promptflow/promptflow/_sdk/_service/utils.py [hunk 4]
```diff
@@ -20,8 +22,26 @@ def wrapper(*args, **kwargs):
...
             return response
 
     return wrapper
+
+
+def validate_request_user():
+    # Get the user name from request.
+    user = request.environ.get("REMOTE_USER") or request.headers.get("X-Remote-User")
+    if user != getpass.getuser():
+        abort(403)
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_service/utils.py:36]: what's the error message user will get in this case?

File: src/promptflow/tests/sdk_pfs_test/utils.py [hunk 5]
```diff
@@ -14,26 +15,35 @@ class PFSOperations:
...
-        return self._client.get(f"{self.RUN_URL_PREFIX}/{name}")
+        return self._client.get(f"{self.RUN_URL_PREFIX}/{name}", headers=self.remote_user_header())
 
     def get_run_metadata(self, name: str):
-        return self._client.get(f"{self.RUN_URL_PREFIX}/{name}/metadata")
+        return self._client.get(f"{self.RUN_URL_PREFIX}/{name}/metadata", headers=self.remote_user_header())
 
     def get_run_detail(self, name: str):
-        return self._client.get(f"{self.RUN_URL_PREFIX}/{name}/detail")
+        return self._client.get(f"{self.RUN_URL_PREFIX}/{name}/detail", headers=self.remote_user_header())
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_pfs_test/utils.py:46]: note we will need to infrom vsc extension on this

File: src/promptflow/promptflow/_sdk/_service/utils.py [hunk 6]
```diff
@@ -20,8 +22,26 @@ def wrapper(*args, **kwargs):
...
         except Exception:  # pylint: disable=broad-except
-            response = jsonify({"error_message": "Internal Server Error"})
+            response = jsonify({"error_message": f"Internal Server Error"})
             response.status_code = 500
             return response
 
     return wrapper
+
+
+def validate_request_user():
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_service/utils.py:32]: note we will need to infrom vsc extension on this. 
Maybe we only add such check for connection calls?

File: src/promptflow/promptflow/_sdk/_constants.py (hunk 1 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/_constants.py:22]: It's used to be hard code in configuration, replaced to constant

File: src/promptflow/promptflow/_sdk/_service/entry.py (hunk 2 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/_service/entry.py:34]: reuse this login in configuration

File: src/promptflow/promptflow/_sdk/_service/entry.py (hunk 3 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/_service/entry.py:42]: write back the random to config yaml

File: src/promptflow/promptflow/_sdk/_service/utils.py (hunk 4 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/_service/utils.py:36]: Response of invalid user
image

File: src/promptflow/promptflow/_sdk/_service/utils.py (hunk 6 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/_service/utils.py:32]: Only connection api will validate the header.

File: src/promptflow/promptflow/_sdk/_service/entry.py [hunk 7]
```diff
@@ -27,19 +27,24 @@ def main():
...
-        # Read and write permission for user.
-        mode = stat.S_IRUSR | stat.S_IWUSR
-        (HOME_PROMPT_FLOW_DIR / SERVICE_CONFIG_FILE).touch(mode=mode, exist_ok=True)
+        (HOME_PROMPT_FLOW_DIR / SERVICE_CONFIG_FILE).touch(mode=read_write_by_user(), exist_ok=True)
         with open(HOME_PROMPT_FLOW_DIR / SERVICE_CONFIG_FILE, "r") as f:
-            service_config = yaml.safe_load(f)
+            service_config = yaml.safe_load(f) or {}
             port = service_config.get("service", {}).get("port", None)
+        if not port:
+            with open(HOME_PROMPT_FLOW_DIR / SERVICE_CONFIG_FILE, "w") as f:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_service/entry.py:38]: maybe we should have another file store the actual port. 
e.g. ~/.promptflow/pfs.port

File: src/promptflow/promptflow/_sdk/_service/utils.py [hunk 8]
```diff
@@ -22,18 +21,22 @@ def wrapper(*args, **kwargs):
...
 
     return wrapper
 
 
-def validate_request_user():
-    # Get the user name from request.
-    user = request.environ.get("REMOTE_USER") or request.headers.get("X-Remote-User")
-    if user != getpass.getuser():
-        abort(403)
+def local_user_only_api_wrapper(func):
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_service/utils.py:31]: don't like the name, wrapper

File: src/promptflow/promptflow/_sdk/_service/entry.py (hunk 7 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/_service/entry.py:38]: change to pfs.port

File: src/promptflow/promptflow/_sdk/_service/utils.py (hunk 8 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/_service/utils.py:31]: change func name to local_user_only

File: src/promptflow/promptflow/_sdk/_service/entry.py [hunk 9]
```diff
@@ -1,16 +1,50 @@
...
+                # Set random port to ~/.promptflow/pf.yaml
+                port = get_random_port()
+                service_config["service"] = service_config.get("service", {})
+                service_config["service"]["port"] = port
+                yaml.dump(service_config, f)
+
     app = create_app()
-    waitress.serve(app, port=LOCAL_SERVICE_PORT)
+    # Set host to localhost, only allow request from localhost.
+    waitress.serve(app, host="127.0.0.1", port=port)
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_service/entry.py:47]: next pr: handle error that already has a service on this port.

Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/_service/entry.py:49]: check whether the port is used

**Review by wangchao1230 — APPROVED** [review]

**Review by crazygao — APPROVED** [review]

**Pull request merged**